src/zmalloc.c

```diff
@@ -361,22 +361,38 @@ size_t zmalloc_usable_size(void *ptr) {
 }
 #endif
 
-void zfree(void *ptr) {
-#ifndef HAVE_MALLOC_SIZE
-    void *realptr;
-    size_t oldsize;
+/* Frees the memory buffer pointed by ptr and updates statistics. When using
+ * jemalloc this function uses the fast track by specifying the buffer size.
+ *
+ * ptr must point to the start of the buffer. On systems where we have
+ * the additional header with size, the caller must do the necessary adjustments
+ * to ptr. ptr must not be NULL.
+ * The caller is responsible to provide the real allocaction size.
+ *
+ * If the caller can't satisfy any of the conditions above, it should call
+ * zfree() instead. */
```

Comment:
```suggestion
/* Frees the memory buffer pointed by ptr and updates statistics. When using
 * jemalloc this function uses the size hint in order free the memory on the fast track.
 * It is the caller responsibility to provide ptr as a pointer to the start of the allocated memory as was previously 
 * returned by the call to allocate the memory (ie. zalloc, zrealloc, ztrymalloc etc...).
 * The size is the real size of the allocated memory. The caller is responsible to provide the real allocation size such  
 * that size == zmalloc_size(ptr).
 *
 * In case the above requirements cannot be met. zfree() can be used instead. */
```

Reply:
I would suggest to refactor (added suggestion) and keep the scope to the z allocation.

Reply:
in fact it does brings up an issue we did not discuss which is the support when HAVE_MALLOC_SIZE is missing. Discussed over offline we agreed that the z* API should encapsulate PREFIX logic, so we will make  zfree_with_size encapsulate that logic.